# Extract Datasets and Establish Benchmark

**Learning Objectives**
- Divide into Train, Evaluation and Test datasets
- Understand why we need each
- Pull data out of BigQuery and into CSV
- Establish Rules Based Benchmark

In the previous notebook we demonstrated how to do ML in BigQuery. However BQML is limited to linear models.

For advanced ML we need to pull the data out of BigQuery and load it into a ML Framework, in our case TensorFlow.

While TensorFlow [can read from BigQuery directly](https://www.tensorflow.org/api_docs/python/tf/contrib/cloud/BigQueryReader), the performance is slow. The best practice is to first stage the BigQuery files as .csv files, and then read the .csv files into TensorFlow. 

The .csv files can reside on local disk if we're training locally, but if we're training in the cloud we'll need to move the .csv files to the cloud, in our case Google Cloud Storage.

## Review

In the [a_sample_explore_clean](a_sample_explore_clean.ipynb) notebook we came up with the following query to extract a repeatable and clean sample: 
<pre>
#standardSQL
SELECT
  (tolls_amount + fare_amount) AS fare_amount, -- label
  pickup_datetime,
  pickup_longitude, 
  pickup_latitude, 
  dropoff_longitude, 
  dropoff_latitude
FROM
  `nyc-tlc.yellow.trips`
WHERE
  -- Clean Data
  trip_distance > 0
  AND passenger_count > 0
  AND fare_amount >= 2.5
  AND pickup_longitude > -78
  AND pickup_longitude < -70
  AND dropoff_longitude > -78
  AND dropoff_longitude < -70
  AND pickup_latitude > 37
  AND pickup_latitude < 45
  AND dropoff_latitude > 37
  AND dropoff_latitude < 45
  -- repeatable 1/5000th sample
  AND MOD(ABS(FARM_FINGERPRINT(CAST(pickup_datetime AS STRING))),5000) = 1
  </pre>
  
We will use the same query **with one change**. Instead of using `pickup_datetime` as is, we will extract `dayofweek` and `hourofday` from it. This is to give us some categorical features in our dataset so we can illustrate how to deal with them when we get to feature engineering. The new query will be:

<pre>
SELECT
  (tolls_amount + fare_amount) AS fare_amount, -- label
  EXTRACT(DAYOFWEEK from pickup_datetime) AS dayofweek,
  EXTRACT(HOUR from pickup_datetime) AS hourofday,
  pickup_longitude, 
  pickup_latitude, 
  dropoff_longitude, 
  dropoff_latitude
-- rest same as before
</pre>

## Split into Train, Evaluation, and Test

For ML modeling we need not just one, but three datasets.

**Train:** This is what our model learns on

**Evaluation (aka Validation):** We shouldn't evaluate our model on the same data we trained on, because then we couldn't know whether it was memorizing the input data or whether it was generalizing. Therefore we evaluate on the evaluation dataset, aka validation dataset.

**Test:** We use our evaluation dataset to tune our hyperparameters (we'll cover hyperparameter tuning in a future lesson). We need to know that our choosen set of hyperparameters will work well for data we haven't seen before, because in production that will be the case. For this reason we create a third dataset that we never use during the model development process. We only evaluate on this once our model development is finished. Data scientists don't always create a test dataset (aka holdout dataset), but to be thorough you should.

We can divide our existing 1/5000th sample three ways 70%/15%/15%  (or whatever split we like) with some modulo math demonstrated below.

Because we are using a hash function these results are deterministic, we'll get the same exact split every time the query is run (assuming the underlying data hasn't changed)

In [1]:
def create_query(phase, sample_size):
  basequery = """
    SELECT
      (tolls_amount + fare_amount) AS fare_amount,
      EXTRACT(DAYOFWEEK from pickup_datetime) AS dayofweek,
      EXTRACT(HOUR from pickup_datetime) AS hourofday,
      pickup_longitude AS pickuplon,
      pickup_latitude AS pickuplat,
      dropoff_longitude AS dropofflon,
      dropoff_latitude AS dropofflat
    FROM
      `nyc-tlc.yellow.trips`
    WHERE
      trip_distance > 0
      AND fare_amount >= 2.5
      AND pickup_longitude > -78
      AND pickup_longitude < -70
      AND dropoff_longitude > -78
      AND dropoff_longitude < -70
      AND pickup_latitude > 37
      AND pickup_latitude < 45
      AND dropoff_latitude > 37
      AND dropoff_latitude < 45
      AND passenger_count > 0
      AND MOD(ABS(FARM_FINGERPRINT(CAST(pickup_datetime AS STRING))), EVERY_N) = 1
  """

  if phase == 'TRAIN':
    subsample = """
    AND MOD(ABS(FARM_FINGERPRINT(CAST(pickup_datetime AS STRING))), EVERY_N * 100) >= (EVERY_N * 0)
    AND MOD(ABS(FARM_FINGERPRINT(CAST(pickup_datetime AS STRING))), EVERY_N * 100) <  (EVERY_N * 70)
    """
  elif phase == 'VALID':
    subsample = """
    AND MOD(ABS(FARM_FINGERPRINT(CAST(pickup_datetime AS STRING))), EVERY_N * 100) >= (EVERY_N * 70)
    AND MOD(ABS(FARM_FINGERPRINT(CAST(pickup_datetime AS STRING))), EVERY_N * 100) <  (EVERY_N * 85)
    """
  elif phase == 'TEST':
    subsample = """
    AND MOD(ABS(FARM_FINGERPRINT(CAST(pickup_datetime AS STRING))), EVERY_N * 100) >= (EVERY_N * 85)
    AND MOD(ABS(FARM_FINGERPRINT(CAST(pickup_datetime AS STRING))), EVERY_N * 100) <  (EVERY_N * 100)
    """
    
  query = basequery + subsample
  return query.replace("EVERY_N", sample_size)

## Write to CSV
Now let's execute a query for train/valid/test and write the results to disk in csv format. We use Pandas's `.to_csv()` method to do so.

In [2]:
import google.datalab.bigquery as bq

for phase in ['TRAIN', 'VALID', 'TEST']:
  # 1. Create query string
  query_string = create_query(phase, '5000')
  # 2. Load results into DataFrame
  df = bq.Query(query_string).execute().result().to_dataframe()
  # 3. Write DataFrame to CSV
  df.to_csv('taxi-{}.csv'.format(phase.lower()), index_label = False, index = False)
  print("Wrote {} lines to {}".format(len(df), 'taxi-{}.csv'.format(phase.lower())))

Wrote 151614 lines to taxi-train.csv
Wrote 31820 lines to taxi-valid.csv
Wrote 34080 lines to taxi-test.csv


Note that even with a 1/5000th sample we have a good amount of data for ML. 150K training examples and 30K validation.

<h3> Verify that datasets exist </h3>

In [3]:
!ls -l *.csv

-rw-r--r-- 1 root root 1799474 Jan 11 14:25 taxi-test.csv
-rw-r--r-- 1 root root 7986353 Jan 11 14:25 taxi-train.csv
-rw-r--r-- 1 root root 1673742 Jan 11 14:25 taxi-valid.csv


### Preview one of the files

In [10]:
!head taxi-train.csv

fare_amount,dayofweek,hourofday,pickuplon,pickuplat,dropofflon,dropofflat
3.8,1,0,-73.986872,40.739698,-73.985832,40.744041
4.0,7,0,-73.97831,40.748493,-73.983726,40.743893
10.5,4,0,-73.990474,40.719493,-73.989566,40.747033
10.5,1,0,-73.990456,40.724912,-73.941419,40.71918
5.3,5,0,-73.974802,40.754947,-73.98479,40.749647
8.1,7,0,-73.999687,40.738595,-73.986888,40.765102
4.1,5,0,-74.00169,40.726052,-74.004438,40.717167
3.7,1,0,-73.913462,40.7652,-73.919733,40.758182
5.5,7,0,-73.982392,40.731307,-73.983375,40.721962


Looks good! We now have our ML datasets and are ready to train ML models, validate them and test them.

## Establish Rules-Based Benchmark

Before we start building complex ML models, it is a good idea to come up with a simple rules based model and use that as a benchmark. After all, there's no point using ML if it can't beat the traditional rules based approach!

Our rule is going to be to divide the mean fare_amount by the mean estimated distance to come up with a rate and use that to predict. 

Recall we can't use the actual `trip_distance` because we won't have that available at prediction time (depends on the route taken), however we do know the users pick up and drop off location so we can use euclidean distance between those coordinates.

In [8]:
import pandas as pd

def euclidean_distance(df):
  return ((df['pickuplat']-df['dropofflat'])**2 + (df['pickuplon']-df['dropofflon'])**2)**.5

def compute_rmse(actual, predicted):
  return (((actual - predicted)**2).mean())**.5

def print_rmse(df, rate, name):
  print("{} RMSE = {}".format(compute_rmse(df['fare_amount'], rate * euclidean_distance(df)), name))

df_train = pd.read_csv('taxi-train.csv')
df_valid = pd.read_csv('taxi-valid.csv')

rate = df_train['fare_amount'].mean() / euclidean_distance(df_train).mean()

print_rmse(df_train, rate, 'Train')
print_rmse(df_valid, rate, 'Valid') 

7.823827497088756 RMSE = Train
7.70415645102671 RMSE = Valid


The simple distance-based rule gives us an RMSE of <b>$7.70</b> on the validation dataset.  We have to beat this, of course, but you will find that simple rules of thumb like this can be surprisingly difficult to beat. 

You don't want to set a goal on the test dataset because you'll want to tweak your hyperparameters and model architecture to get the best validation error. Then, you can evaluate ONCE on the test data.

## Challenge Exercise

Let's say that you want to predict whether a Stackoverflow question will be acceptably answered. Using this [public dataset of questions](https://bigquery.cloud.google.com/table/bigquery-public-data:stackoverflow.posts_questions), create a machine learning dataset that you can use for classification.
<p>
What is a reasonable benchmark for this problem?
What features might be useful?
<p>
If you got the above easily, try this harder problem: you want to predict whether a question will be acceptably answered within 2 days. How would you create the dataset?
<p>
Hint (highlight to see):
<p style='color:white' linkstyle='color:white'> 
You will need to do a SQL join with the table of [answers]( https://bigquery.cloud.google.com/table/bigquery-public-data:stackoverflow.posts_answers) to determine whether the answer was within 2 days.
</p>

Copyright 2019 Google Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.